plugin/rename_file.py

```diff
@@ -0,0 +1,149 @@
+        old_path = paths[0] if paths else None
+        if old_path:
+            self.window.run_command('lsp_rename_path', {
+                "old_path": old_path
+            })
+
+
+class RenamePathInputHandler(sublime_plugin.TextInputHandler):
+    def __init__(self, path: str) -> None:
+        self.path = path
+
+    def name(self) -> str:
+        return "new_name"
+
+    def placeholder(self) -> str:
+        return self.path
+
+    def initial_text(self) -> str:
+        return self.placeholder()
+
+    def initial_selection(self) -> list[tuple[int, int]]:
+        name = Path(self.path).stem
+        return [(0, len(name))]
+
+    def validate(self, path: str) -> bool:
+        return len(path) > 0
+
+
+class LspRenamePathCommand(LspWindowCommand):
+    capability = 'workspace.fileOperations.willRename'
+
+    def is_enabled(self) -> bool:
+        return True
+
+    def want_event(self) -> bool:
+        return False
+
+    def input(self, args: dict) -> sublime_plugin.TextInputHandler | None:
+        if "new_name" in args:
+            return None
+        view = self.window.active_view()
+        old_path = args.get('old_path') or view.file_name() if view else None
+        return RenamePathInputHandler(Path(old_path or "").name)
+
+    def run(self, new_name: str, old_path: str | None = None) -> None:
+        view = self.window.active_view()
+        old_path = old_path or view.file_name() if view else None
+        if old_path is None:  # handle renaming buffers
+            if view:
+                view.set_name(new_name)
+            return
+        if new_name == old_path:
+            return
+        # new_name can be: FILE_NAME.xy OR ./FILE_NAME.xy OR ../../FILE_NAME.xy
+        resolved_new_path = (Path(old_path).parent / new_name).resolve()
+        new_path = str(resolved_new_path)
+        if resolved_new_path.exists() and not self.is_case_change(old_path, new_path):
+            self.window.status_message('Unable to Rename. Already exists')
+            return
+        file_rename: FileRename = {
+            "newUri": filename_to_uri(new_path),
+            "oldUri": filename_to_uri(old_path)
+        }
+
+        def create_request_async(session: Session) -> Promise[tuple[WorkspaceEdit | None, weakref.ref[Session]]]:
+            return session.send_request_task(
+                        Request.willRenameFiles({'files': [file_rename]})
+                   ).then(lambda response: (response, weakref.ref(session)))
+
+        sessions = [session for session in self.sessions() if match_file_operation_filters(
+            session.get_capability('workspace.fileOperations.willRename'), file_rename['oldUri']
+        )]
+        promises = [create_request_async(session) for session in sessions]
+        if promises:
+            sublime.set_timeout_async(
+                lambda: Promise.all(promises).then(lambda responses: self.handle_rename_async(responses, file_rename))
+            )
```

Comment:
This feels quite wrong. If the idea is that the requests should be made on async thread then this code doesn't do that.

If this actually does make a difference (and I believe if you say so) then there is probably some weird timing issue involved but I would then try a different approach. Create a method like `run_async`, move all the code from the line with `file_rename: FileRename = {` into it and call it using `sublime.set_timeout_async` from `run`. Then the requests will actually be made from async thread and it might also fix whatever issue you are seeing.

Reply:
[cfa5799](https://github.com/sublimelsp/LSP/pull/2498/commits/cfa5799d3ff91d6a5056d0411a4f8a8342025b27)

Reply:
And remove `sublime.set_timeout_async(` from the line that I've highlighted.